students/Dustin_L/lesson04/mailroom2.py

```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env python3
+"""Mail Room 2 Module
+
+This module contains all of the functions for the updated Mail Room 2 module.
+"""
+
+from collections import defaultdict
+import datetime
+import copy
+
+THANK_YOU_OPT = 1
+REPORT_OPT = 2
+LETTERS_OPT = 3
+QUIT_OPT = 4
+
+GIFTS_KEY = 'Gifts'
+NUM_GIFTS_KEY = 'Number of Gifts'
+TOTAL_KEY = 'Total'
+AVE_KEY = 'Average'
+DEFAULT_DICT = {GIFTS_KEY: [],
+                NUM_GIFTS_KEY: 0,
+                TOTAL_KEY: 0,
+                AVE_KEY: 0}
+DONOR_DB = defaultdict(lambda: copy.deepcopy(DEFAULT_DICT),
+                       {'Toni Morrison': {GIFTS_KEY: [1000, 5000, 10000],
+                                          NUM_GIFTS_KEY: 0,
+                                          TOTAL_KEY: 0,
+                                          AVE_KEY: 0},
+                        'Mike McHargue': {GIFTS_KEY: [12000, 50000, 27000],
+                                          NUM_GIFTS_KEY: 0,
+                                          TOTAL_KEY: 0,
+                                          AVE_KEY: 0},
+                        "Flannery O'Connor": {GIFTS_KEY: [38734, 6273, 67520],
+                                              NUM_GIFTS_KEY: 0,
+                                              TOTAL_KEY: 0,
+                                              AVE_KEY: 0},
+                        'Angela Davis': {GIFTS_KEY: [74846, 38470, 7570, 50],
+                                         NUM_GIFTS_KEY: 0,
+                                         TOTAL_KEY: 0,
+                                         AVE_KEY: 0},
+                        'Bell Hooks': {GIFTS_KEY: [634547, 47498, 474729, 4567],
+                                       NUM_GIFTS_KEY: 0,
+                                       TOTAL_KEY: 0,
+                                       AVE_KEY: 0}})
```

Comment:
you were definitely on the right track here but I do have couple comments:
* `DONOR_DB` is not a constant (constants don't change), so in this case it should be a lower case global var
* using constants as key name definitions isn't very pythonic although I can see potential benefits
* dictionary keys are _usually_ concise name using camel case or snake case. If you're returning JSON structure it is more common to use camelCase (don't worry if you don't know what JSON is), otherwise it's snake_case.

here is altered version considering my above points:
```
>>> def initialize_donor_data(gifts=None):  # you could add keywords for other data (total etc) but not our use case for now
...     return {"gifts": gifts or [],
...             "num_of_gifts": 0,
...             "total": 0,
...             "avg": 0}
...
>>>
>>> donor_db = defaultdict(lambda: initialize_donor_data(),
...                        {'Toni Morrison': initialize_donor_data(gifts=[1000, 5000, 10000]),
...                         'Mike McHargue': initialize_donor_data(gifts=[12000, 50000, 27000]),
...                         "Flannery O'Connor": initialize_donor_data(gifts=[38734, 6273, 67520]),
...                         'Angela Davis': initialize_donor_data(gifts=[74846, 38470, 7570, 50]),
...                         'Bell Hooks': initialize_donor_data(gifts=[634547, 47498, 474729, 4567])})



```

Reply:
I like this approach, however, it seems that using the function call in the dict constructor results in the function getting called when the key is selected. When I test ran this, I hit an exception when the add_donation() function was called and it tried to update the database:
File "./mailroom2.py", line 76, in add_donation
    donor_db[donor][GIFTS_KEY].append(amount)
TypeError: 'function' object is not subscriptable

Is there a way to tell python to use the function call only on instantiation of the dict instead of it being invoked with each key access (like is done with opt_dict)?

Reply:
@rdlaner I tried to reproduce your error but everything is working fine:
```
from collections import defaultdict


def initialize_donor_data(gifts=None):  # you could add keywords for other data (total etc) but not our use case for now
     return {"gifts": gifts or [],
             "num_of_gifts": 0,
             "total": 0,
             "avg": 0}


donor_db = defaultdict(lambda: initialize_donor_data(),
                        {'Toni Morrison': initialize_donor_data(gifts=[1000, 5000, 10000]),
                         'Mike McHargue': initialize_donor_data(gifts=[12000, 50000, 27000]),
                         "Flannery O'Connor": initialize_donor_data(gifts=[38734, 6273, 67520]),
                         'Angela Davis': initialize_donor_data(gifts=[74846, 38470, 7570, 50]),
                         'Bell Hooks': initialize_donor_data(gifts=[634547, 47498, 474729, 4567])})

print(donor_db)


donor = "Bell Hooks"
amount = 10
donor_db[donor]["gifts"].append(amount)

donor_db["Natasha"]["gifts"].append(amount)

print(donor_db)
```

if you provide the code that gave you this error I can help you debug, doesn't have to be a PR - you can just push your code to your fork and let me know so I can run it.

Reply:
Hmmm. Ok, I pushed up my latest changes to my fork. Let me know what you think. I run into the exception after I attempt to send a thank you to a person not in the database.

Reply:
I can confirm that it works fine when accessing a member that is already in the dictionary, but when trying to add a key that isn't already in the dictionary, it seems that defaultdict is having an issue with the init_donor_data() function.

Reply:
Ah! I think I found my error. I had left off the parenthesis for the init_donor_data function when called from the lambda on line 41.

Reply:
yay! glad you got it figured out 🙂